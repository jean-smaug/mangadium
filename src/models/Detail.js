import { types, flow, getParent } from 'mobx-state-tree'
import { bool, shape } from 'prop-types'

import { MangaDetail, MangaDetailPropTypes } from './Manga'
import request from '../utils/request'

export const DetailPropTypes = shape({
  manga: MangaDetailPropTypes,
  isOpen: bool.isRequired
})

export const Detail = types
  .model({
    manga: types.maybeNull(MangaDetail),
    isOpen: types.optional(types.boolean, false)
  })
  .actions(self => ({
    toggleVisibilityStatus () {
      self.isOpen = !self.isOpen
    },
    setManga: flow(function * (id) {
      const appStore = getParent(self).app

      try {
        appStore.toggleLoadingStatus()

        const manga = yield request.getMangaAndPictures(id)
        self.manga = manga

        appStore.toggleLoadingStatus()
        self.toggleVisibilityStatus()
      } catch (error) {
        console.error(error)
        appStore.toggleLoadingStatus()
      }
    })
  }))
