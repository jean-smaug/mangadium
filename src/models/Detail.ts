import { types, flow, getParent } from 'mobx-state-tree'
import { bool } from 'prop-types'

import { MangaDetail, MangaDetailInterface } from './Manga'
import request from '../utils/request'

export interface DetailInterface {
  manga: MangaDetailInterface,
  isOpen: boolean,
  toggleVisibilityStatus: () => void,
  setManga: (id: number) => void
}

export const Detail = types
  .model({
    manga: types.maybeNull(MangaDetail),
    isOpen: types.optional(types.boolean, false)
  })
  .actions(self => ({
    toggleVisibilityStatus () {
      self.isOpen = !self.isOpen
    },
    setManga: flow(function * (id) {
      const appStore = getParent(self).app

      try {
        appStore.toggleLoadingStatus()

        const manga = yield request.getMangaAndPictures(id)
        self.manga = manga

        appStore.toggleLoadingStatus()
        self.toggleVisibilityStatus()
      } catch (error) {
        console.error(error)
        appStore.toggleLoadingStatus()
      }
    })
  }))
