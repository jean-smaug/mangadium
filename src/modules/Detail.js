import React, { Component, Fragment } from "react";
import { observer, inject } from "mobx-react";

import ViewToggler from "../components/ViewToggler";

@inject("detail")
@observer
class Detail extends Component {
  render() {
    const {
      detail,
      detail: { manga }
    } = this.props;

    return (
      <ViewToggler
        isVisible={detail.isOpen}
        onClose={detail.toggleVisibilityStatus}
      >
        <div className="container">
          {manga && (
            <div>
              <h2 className="title is-2">{manga.title}</h2>
              <div className="content">
                <p>{manga.synopsis}</p>
              </div>

              <div className="columns" style={{ overflowX: "scroll" }}>
                {manga.images.map(image => (
                  <div key={image} className="column is-one-quarter">
                    <figure className="image is-3by5">
                      <img src={image} />
                    </figure>
                  </div>
                ))}
              </div>
            </div>
          )}
        </div>
      </ViewToggler>
    );
  }
}

export default Detail;
