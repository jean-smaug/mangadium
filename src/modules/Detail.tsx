import * as React from 'react'
import { observer, inject } from 'mobx-react'

import ViewToggler from '../components/ViewToggler'
import ImageSlider from '../components/ImageSlider'
import { displayAuthors, convertSpecialCharsToText } from '../utils/render'

import { DetailInterface } from '../models/Detail'

interface Props {
  detail: DetailInterface
}
@inject('detail')
@observer
class Detail extends React.Component<Props> {
  render () {
    const {
      detail,
      detail: { manga }
    } = this.props

    return (
      <ViewToggler isVisible={detail.isOpen} onClose={detail.toggleVisibilityStatus}>
        <div className='container'>
          {manga && (
            <div>
              <div style={{ display: 'flex' }}>
                <h2 className='title is-2'>
                  <a href={manga.url} target='_blank'>
                    {manga.title}
                  </a>
                </h2>
              </div>
              <p className='subtitle'>{displayAuthors(manga.authors)}</p>
              <div className='content'>
                {manga.genres.map(genre => (
                  <span key={genre} className='tag is-dark' style={{ marginLeft: '3px' }}>
                    {genre}
                  </span>
                ))}
              </div>
              <div className='content'>
                <p>{convertSpecialCharsToText(manga.synopsis)}</p>
              </div>
              <ImageSlider images={manga.images} />
            </div>
          )}
        </div>
      </ViewToggler>
    )
  }
}

export default Detail
