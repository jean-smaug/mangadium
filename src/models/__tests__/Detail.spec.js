import { Detail } from '../Detail'

describe('<=== Detail ===>', () => {
  let detail = null

  beforeEach(() => {
    detail = Detail.create()
  })

  it('should create store', () => {
    expect(detail).toMatchSnapshot()
  })
})
