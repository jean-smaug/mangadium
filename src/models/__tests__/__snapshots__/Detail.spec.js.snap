// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<=== Detail ===> should create store 1`] = `
Object {
  "isOpen": false,
  "manga": null,
}
`;
